### PR TITLE
[feature](nereids) Matiarilzed view query rewrite util implementation

### PR DESCRIPTION
**Pull request by seawinde**
## Proposed changes

The basic util implementatation which is used by materialized view rewrite

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by seawinde** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/PredicatesSplitter.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    /**PredicateExtract*/
+    public class PredicateExtract extends DefaultExpressionVisitor<Void, Expression> {
+
+        @Override
+        public Void visitComparisonPredicate(ComparisonPredicate comparisonPredicate, Expression sourceExpression) {
+            Expression leftArg = comparisonPredicate.getArgument(0);
+            Expression rightArg = comparisonPredicate.getArgument(1);
+            boolean leftArgOnlyContainsColumnRef = containOnlyColumnRef(leftArg, true);
+            boolean rightArgOnlyContainsColumnRef = containOnlyColumnRef(rightArg, true);
+            if (comparisonPredicate instanceof EqualTo || comparisonPredicate instanceof NullSafeEqual) {
```
Comment (XieJiann) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/PredicatesSplitter.java:66]:
```suggestion
            if (comparisonPredicate instanceof EqualPredicate) {
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/PredicatesSplitter.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+ * Split the expression to equal, range and residual predicate.
+ * Should instance when used.
+ * TODO support complex predicate split
+ */
+public class PredicatesSplitter {
+
+    private List<Expression> equalPredicates = new ArrayList<>();
+    private List<Expression> rangePredicates = new ArrayList<>();
+    private List<Expression> residualPredicates = new ArrayList<>();
+    private List<Expression> conjunctExpressions;
```
Comment (XieJiann) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/PredicatesSplitter.java:46]: final?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/PredicatesSplitter.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        @Override
+        public Void visitComparisonPredicate(ComparisonPredicate comparisonPredicate, Expression sourceExpression) {
+            Expression leftArg = comparisonPredicate.getArgument(0);
+            Expression rightArg = comparisonPredicate.getArgument(1);
+            boolean leftArgOnlyContainsColumnRef = containOnlyColumnRef(leftArg, true);
+            boolean rightArgOnlyContainsColumnRef = containOnlyColumnRef(rightArg, true);
+            if (comparisonPredicate instanceof EqualTo || comparisonPredicate instanceof NullSafeEqual) {
+                if (leftArgOnlyContainsColumnRef && rightArgOnlyContainsColumnRef) {
+                    equalPredicates.add(comparisonPredicate);
+                    return null;
```
Comment (XieJiann) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/PredicatesSplitter.java:69]: redundant

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/PredicatesSplitter.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+            return null;
+        }
+
+        @Override
+        public Void visitCompoundPredicate(CompoundPredicate compoundPredicate, Expression context) {
+            if (compoundPredicate instanceof Or) {
+                residualPredicates.add(compoundPredicate);
+                return null;
+            }
+            return super.visitCompoundPredicate(compoundPredicate, context);
```
Comment (XieJiann) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/PredicatesSplitter.java:88]: It seems should visit their children.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/mapping/RelationMapping.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+                .collect(ImmutableList.toImmutableList());
+    }
+
+    private static String getTableQualifier(TableIf tableIf) {
+        String tableName = tableIf.getName();
+        DatabaseIf database = tableIf.getDatabase();
+        if (database == null) {
+            return null;
+        }
+        return database.getFullName() + ":" + tableName;
```
Comment (XieJiann) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/mapping/RelationMapping.java:124]: Use tableId may be better. It's globally unique

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/mapping/RelationMapping.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+ * such as query pattern is a1 left join a2 left join b
+ * view pattern is a1 left join a2 left join b. the mapping will be
+ * [{a1:a1, a2:a2, b:b}, {a1:a2, a2:a1, b:b}]
+ */
+public class RelationMapping extends Mapping {
+
+    private final BiMap<MappedRelation, MappedRelation> mappedRelationMap;
+
+    public RelationMapping(BiMap<MappedRelation, MappedRelation> mappedRelationMap) {
+        this.mappedRelationMap = mappedRelationMap;
```
Comment (XieJiann) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/mapping/RelationMapping.java:46]: Can this be changed if just passing reference? Maybe it's better to use ImmuatbleBiMap.copyof

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/PredicatesSplitter.java (hunk 3 above)
Comment (seawinde) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/PredicatesSplitter.java:69]: have fixed it

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/PredicatesSplitter.java (hunk 2 above)
Comment (seawinde) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/PredicatesSplitter.java:46]: OK

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/PredicatesSplitter.java (hunk 1 above)
Comment (seawinde) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/PredicatesSplitter.java:66]: have fixed it

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/Predicates.java [hunk 7]
```diff
@@ -74,15 +72,15 @@ public SplitPredicate(Expression equalPredicates, Expression rangePredicates, Ex
             this.residualPredicates = residualPredicates;
         }
 
-        public Expression getEqualPredicates() {
+        public Expression getEqualPredicate() {
             return equalPredicates;
```
[the remark this answers is not in the document]
Comment (seawinde) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/Predicates.java:76]: have fixed it

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/mapping/RelationMapping.java (hunk 5 above)
Comment (seawinde) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/mapping/RelationMapping.java]: OK, you are right, have fixed it

**Review by keanji-x — APPROVED** [review]

**Review by morrySnow — APPROVED** [review]

**Pull request merged**